### PR TITLE
[Bug] Order creation screen freezes after adding products

### LINKED ISSUE #12490: Adding a product in the order creation on phones lags/freezes for 1 second
**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.
* Adding a product in the order creation on phones lags/freezes for 1 second.
* Adding more products increases the delay even more.

Discussion: p1725284759189029-slack-C025A8VV728

**To Reproduce**
1. Create new order
2. Open product selector and add a few products
3. Notice the network call-related delay before the order is updated

**Screenshots**

https://github.com/user-attachments/assets/91b92857-a2f1-440d-a486-838cca926c07

**Expected behavior**
Adding products to the order should not cause a delay.

**Mobile Environment**
Please include:
- Device: pixel 7 pro
- Android version: 14
- WooCommerce Android version: 20.1

**WordPress Environment**
<details>
```
Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin.
```
</details>

### PR DESCRIPTION
**Pull request by samiuelson**
### [Bug] Order creation screen freezes after adding products

Don't merge until the base is merged into trunk.

Closes: #12490

### Description

After adding new products to the order on the Product Selector screen, there was a call to the backend made to fetch a product object for each of the newly selected items.

This PR improves that behavior by checking for the product in the local Products cache, and querying the backend only in case the product doesn't exist locally.

### Steps to reproduce

Add products to the new order and observe the app freezing before updating the order.

### Testing information

Tests should cover the following scenarios:
1. Adding product by scanning the barcode (SKU) in case the products cache is empty (e.g. fresh install, cleared data)
2. Adding products via product selector should not cause any delay

### The tests that have been performed

I've tested the above scenarios on Pixel 7 pro, Android 14.

### Images/gif

|Before|After|
|---|---|
|<video src=https://github.com/user-attachments/assets/8caa27f5-caae-443d-896d-d9dbd27cab98/>|<video src=https://github.com/user-attachments/assets/ff2e2e43-ab9a-40db-bb39-8210b7737024/>|

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/CreateOrderItem.kt [hunk 1]
```diff
@@ -26,7 +26,8 @@ class CreateOrderItem @Inject constructor(
         productConfiguration: ProductConfiguration? = null,
     ): Order.Item {
         return withContext(coroutineDispatchers.io) {
-            val product = productDetailRepository.fetchProductOrLoadFromCache(remoteProductId)
+            val product = productDetailRepository.getProduct(remoteProductId)
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/CreateOrderItem.kt:29]: The fix itself is here

Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/CreateOrderItem.kt:29]: ❓ nitpick:  What do you think about writing unit test for this class? Especially focusing on the newly added logic.

**Review by AnirudhBhat — COMMENTED** [review]
Thanks for working on this @samiuelson. The fix works fine.

Even with this fix, it seems the screen may still appear frozen when products are not found in the cache, as it’s making an API call in the background. To improve the user experience, I suggest showing a visual progress indicator during this process, so users are aware that something is happening.

`viewState = viewState.copy(isUpdatingOrderDraft = true|false)`

Let me know what you think!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/CreateOrderItem.kt (hunk 1 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/CreateOrderItem.kt:29]: Added tests: [430bcbe](https://github.com/woocommerce/woocommerce-android/pull/12493/commits/430bcbe2f5d0b3e549fb4a9852c17458c187539b)

**Comment by samiuelson** [comment]
Thanks for the thorough review, @AnirudhBhat! Nice enhancement ideas, I applied them in the following commits:
* [Show loading indicator while computing selected items diff](https://github.com/woocommerce/woocommerce-android/pull/12493/commits/11a0b369db51223074721cbfb5f5f408731249d4) - [11a0b36](https://github.com/woocommerce/woocommerce-android/pull/12493/commits/11a0b369db51223074721cbfb5f5f408731249d4)
* [Add unit tests](https://github.com/woocommerce/woocommerce-android/pull/12493/commits/430bcbe2f5d0b3e549fb4a9852c17458c187539b) - [430bcbe](https://github.com/woocommerce/woocommerce-android/pull/12493/commits/430bcbe2f5d0b3e549fb4a9852c17458c187539b)
Let me know what you think!

**Review by AnirudhBhat — APPROVED** [review]
LGTM 🚀

**Pull request merged**